crates/engine/src/response/write/subgraph_response.rs

```diff
@@ -0,0 +1,166 @@
+use std::{
+    cell::{Ref, RefCell, RefMut},
+    rc::Rc,
+    sync::Arc,
+};
+
+use crate::{
+    execution::ExecutionContext,
+    operation::ResponseObjectSetDefinitionId,
+    response::{
+        ConcreteShapeId, DataPart, GraphqlError, InputResponseObjectSet, ResponseObjectField, ResponseObjectRef,
+        ResponseObjectSet, ResponseValueId,
+    },
+    Runtime,
+};
+
+use super::deserialize::UpdateSeed;
+
+pub(crate) struct SubgraphResponse {
+    pub(super) data: DataPart,
+    shape_id: ConcreteShapeId,
+    pub(super) root_response_object_set: Arc<InputResponseObjectSet>,
+    pub(super) propagated_null_up_to_root: bool,
+    pub(super) propagated_null_up_to_paths: Vec<Vec<ResponseValueId>>,
+    pub(super) errors: Vec<GraphqlError>,
+    pub(super) subgraph_errors: Vec<GraphqlError>,
+    pub(super) updates: Vec<ObjectUpdate>,
+    pub(super) response_object_sets: Vec<(ResponseObjectSetDefinitionId, ResponseObjectSet)>,
+}
+
+impl SubgraphResponse {
+    pub(super) fn new(
+        data: DataPart,
+        shape_id: ConcreteShapeId,
+        root_response_object_set: Arc<InputResponseObjectSet>,
+    ) -> Self {
+        Self {
+            data,
+            shape_id,
+            root_response_object_set,
+            propagated_null_up_to_root: false,
+            propagated_null_up_to_paths: Vec::new(),
+            errors: Vec::new(),
+            subgraph_errors: Vec::new(),
+            updates: Vec::new(),
+            response_object_sets: Vec::new(),
+        }
+    }
+
+    /// Executors manipulate the response within a Send future, so we can't use a Rc/RefCell
+    /// directly. Only once the executor is ready to write should it use this method.
+    pub fn as_mut(&mut self) -> SubgraphResponseRefMut<'_> {
+        SubgraphResponseRefMut {
+            inner: Rc::new(RefCell::new(self)),
+        }
+    }
+
+    fn propagate_null(&mut self, path: &[ResponseValueId]) {
+        let Some(i) = path.iter().rev().position(|value| value.is_nullable()) else {
+            self.propagated_null_up_to_root = true;
+            return;
+        };
+        // we inverted the path.
+        let i = path.len() - i - 1;
+
+        // if let Some(value_id) = path.get(i).filter(|value_id| value_id.part_id() == self.data.id) {
+        //     self.data.make_inaccessible(*value_id);
+        // } else {
+        self.propagated_null_up_to_paths.push(path[..(i + 1)].to_vec());
+        // }
```

Comment:
What's with the commented out code?

Reply:
ah right forgot it, if the value is within the current part, I can just make the value inaccessible immediately. But somehow had an issue with it initially, so commented out to then fix further problems. bringing it back or giving up on it. thanks